cmd/car/extract.go

```diff
@@ -0,0 +1,226 @@
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-unixfsnode"
+	"github.com/ipfs/go-unixfsnode/data"
+	"github.com/ipfs/go-unixfsnode/file"
+	"github.com/ipld/go-car/v2/blockstore"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
+	"github.com/urfave/cli/v2"
+)
+
+// ExtractCar pulls files and directories out of a car
+func ExtractCar(c *cli.Context) error {
+	outputDir, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	if c.Args().Present() {
+		outputDir = c.Args().First()
+	}
+
+	bs, err := blockstore.OpenReadOnly(c.String("file"))
+	if err != nil {
+		return err
+	}
+
+	ls := cidlink.DefaultLinkSystem()
+	ls.TrustedStorage = true
+	ls.StorageReadOpener = func(_ ipld.LinkContext, l ipld.Link) (io.Reader, error) {
+		cl, ok := l.(cidlink.Link)
+		if !ok {
+			return nil, fmt.Errorf("not a cidlink")
+		}
+		blk, err := bs.Get(cl.Cid)
+		if err != nil {
+			return nil, err
+		}
+		return bytes.NewBuffer(blk.RawData()), nil
+	}
+
+	roots, err := bs.Roots()
+	if err != nil {
+		return err
+	}
+
+	for _, root := range roots {
+		if err := extractRoot(c, &ls, root, outputDir); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func extractRoot(c *cli.Context, ls *ipld.LinkSystem, root cid.Cid, outputDir string) error {
+	if root.Prefix().Codec == cid.Raw {
+		if c.IsSet("verbose") {
+			fmt.Fprintf(c.App.ErrWriter, "skipping raw root %s\n", root)
+		}
+		return nil
+	}
+
+	pbn, err := ls.Load(ipld.LinkContext{}, cidlink.Link{Cid: root}, dagpb.Type.PBNode)
```

Comment:
This code seems to assume that if you don't have a raw node, you have a root that is a UnixFS Directory, not a regular file (even though extractFile exists as a function). Is this the intent?

Reply:
I believe this is generally the case. if i export a single file it'll be implicitly in a directory in order to provide the file name of the file.

Reply:
fwiw, to compare to other unix stuff, `tar` commands do allow single file no-dir contents, iirc.  Rare and a bit arcane, but possible.

Reply:
okay for that to go in a follow-up?